theano/tensor/nnet/spatialtf.py

```diff
@@ -0,0 +1,413 @@
+            x1 = np.minimum(x1_f % w, -x1_f % w)
+            h = 2 * (height_f - 1)
+            y0 = np.minimum(y0_f % h, -y0_f % h)
+            y1 = np.minimum(y1_f % h, -y1_f % h)
+        elif border_mode == 'wrap':
+            x0 = np.mod(x0_f, width_f)
+            x1 = np.mod(x1_f, width_f)
+            y0 = np.mod(y0_f, height_f)
+            y1 = np.mod(y1_f, height_f)
+        else:
+            raise ValueError("border_mode must be one of "
+                             "'nearest', 'mirror', 'wrap'")
+        x0, x1, y0, y1 = np.asarray([x0, x1, y0, y1]).astype('int64')
+
+        # Compute offsets and indices
+        height_stride = width
+        batch_stride = width * height
+        base = np.repeat(np.arange(num_batch, dtype='int64') * batch_stride,
+                         out_height * out_width)
+        base_y0 = base + y0 * height_stride
+        base_y1 = base + y1 * height_stride
+        idx_a = base_y0 + x0
+        idx_b = base_y1 + x0
+        idx_c = base_y0 + x1
+        idx_d = base_y1 + x1
+        # Use indices to lookup pixels for all samples
+        inp_flat = inp_transposed.reshape((-1, num_channels))
+        Ia = inp_flat[idx_a]
+        Ib = inp_flat[idx_b]
+        Ic = inp_flat[idx_c]
+        Id = inp_flat[idx_d]
+        # Compute bilinear interpolation weights
+        wa = ((x1_f - x) * (y1_f - y))[:, np.newaxis]
+        wb = ((x1_f - x) * (y - y0_f))[:, np.newaxis]
+        wc = ((x - x0_f) * (y1_f - y))[:, np.newaxis]
+        wd = ((x - x0_f) * (y - y0_f))[:, np.newaxis]
+        # Compute interpolated values
+        transformed_inputs = np.sum([wa * Ia, wb * Ib, wc * Ic, wd * Id], axis=0)
+        # Reshape flat array into NHWC format
+        output = np.reshape(transformed_inputs, (num_batch, out_height, out_width, num_channels))
+        # Dimshuffle NHWC tensor into NCHW format
+        out[0] = np.transpose(output, axes=(0, 3, 1, 2)).astype(inp.dtype)
+
+
+class TransformerGradI(Op):
+    def __init__(self, border_mode):
+        self.border_mode = border_mode
+
+    def make_node(self, inp, grid, grad_outputs):
+        _inp = cpu_contiguous(as_tensor_variable(inp))
+        assert _inp.ndim == 4
+
+        _grid = cpu_contiguous(as_tensor_variable(grid))
+        assert _grid.ndim == 4
+
+        _grad_outputs = cpu_contiguous(as_tensor_variable(grad_outputs))
+        assert _grad_outputs.ndim == 4
+
+        grad_inp = theano.tensor.tensor(dtype=_inp.type.dtype,
+                                        broadcastable=_inp.broadcastable)
+
+        grad_grid = theano.tensor.tensor(dtype=_inp.type.dtype,
+                                         broadcastable=_inp.broadcastable)
+
+        return Apply(self, [_inp, _grid, _grad_outputs], [grad_inp, grad_grid])
+
+    def perform(self, node, inputs, output_storage):
+        inp, grid, grad_outputs = inputs
+        assert len(inp.shape) == 4
+        assert len(grid.shape) == 4
+        assert len(grad_outputs.shape) == 4
```

Comment:
```python
assert inp.ndim == 4
assert grid.ndim == 4
assert grad_outputs.ndim == 4
```
But, as it is already checked in make_node(), you don't need to check it here again. Consider removing these checkings here.